### PR TITLE
Enlighten AssignProjectConfiguration for multithreaded mode

### LINKED ISSUE #13613: Enlighten AssignProjectConfiguration task for multithreaded mode
# Enlighten AssignProjectConfiguration task for multithreaded mode

Parent: #11834

## Context

`AssignProjectConfiguration` is the **highest-execution unmigrated task** (22.9B total executions) listed in the migration epic (#11834) under "Other (either simple or with unknown issues)". It maps solution configuration/platform pairs onto project references, optionally adding synthetic project references for solution-level dependencies.

The task has **no cwd-dependent operations**:
- No `Environment.CurrentDirectory`, `Path.GetFullPath`, or `File.*`/`Directory.*` usage
- All path resolution uses MSBuild's `GetMetadata("FullPath")` which returns absolute paths resolved by the engine
- Platform mapping dictionaries are instance-scoped lazy-init — no cross-thread sharing
- The only static method (`SetBuildInProjectAndReferenceOutputAssemblyMetadata`) is pure logic with no mutable state
- Base class `ResolveProjectBase` uses `reference.GetMetadata("FullPath")` — engine-resolved absolute paths

This is an **attribute-only** migration: the task needs `[MSBuildMultiThreadableTask]` but does NOT need `IMultiThreadableTask` or `TaskEnvironment` because it performs no file system operations, environment variable access, or process launches.

## Approach

Apply the **PR #13045 (attribute-only) pattern**: add `[MSBuildMultiThreadableTask]` to the class declaration. No `IMultiThreadableTask` implementation is needed since the task has no cwd-dependent operations.

1. Decorate `AssignProjectConfiguration` with `[MSBuildMultiThreadableTask]`.
2. No other code changes required — the task is already thread-safe by design.

## ChangeWave consideration

No behavioral change — the attribute simply declares the task safe for multithreaded scheduling. Observable behavior (return value, `[Output]` properties, error/warning messages) is byte-identical. No ChangeWave needed.

### Existing unit tests (`src/Tasks.UnitTests/AssignProjectConfiguration_Tests.cs`)

Existing tests cover configuration assignment, platform mapping, synthetic project references, and edge cases. Coverage is good for the functional behavior.

### Gaps to fill in this PR

- **G1 — Concurrency test.** Two `AssignProjectConfiguration` instances with different `SolutionConfigurationContents` and `ProjectReferences` execute concurrently, asserting each produces correct, independent output. Verifies no shared mutable state interference.
- **G2 — Baseline parity.** Verify all existing tests continue to pass with no behavioral changes. `[Output] AssignedProjects` items have identical metadata (Configuration, Platform, SetConfiguration, SetPlatform, etc.) before and after the attribute addition.

## Acceptance criteria

- [ ] `AssignProjectConfiguration` decorated with `[MSBuildMultiThreadableTask]`.
- [ ] No `IMultiThreadableTask` implementation (not needed — no cwd-dependent operations).
- [ ] All existing `AssignProjectConfiguration_Tests` pass on `net472` and `net10.0`.
- [ ] `[Output] AssignedProjects` items have byte-identical `ItemSpec` and metadata.
- [ ] Concurrency test (**G1**) added and passes.
- [ ] No new compiler warnings (warnings-as-errors in official build).
- [ ] `multithreaded-task-migration` SKILL sign-off checklist walked and passes.

## Implementation order

1. Apply `[MSBuildMultiThreadableTask]` to `AssignProjectConfiguration`.
2. Add **G1** (concurrency) test.
3. Run `Tasks.UnitTests`; verify clean.
4. Run repo build with `-v quiet` to ensure no new warnings.

## Risks / open questions

- Base class `ResolveProjectBase` is also used by `ResolveNonMSBuildProjectOutput`. The attribute has `Inherited = false`, so each concrete class must be independently decorated — no implicit inheritance of the attribute from the base class. This is correct and intentional.
- `AddSyntheticProjectReferences` mutates the `_projectReferences` array via `ResolveProjectBase`. This is instance state, not static — safe under the per-instance-per-thread model.

## References

- Parent epic: #11834
- Pattern PRs:
  - #13045 — Enlighten more tasks that require no change (attribute-only baseline)
  - #13267 — Enlighten public versions of intrinsic tasks
  - #13069 — Enlighten FindUnderPath and AssignTargetPath
- SKILL: `.github/skills/multithreaded-task-migration/SKILL.md`

### PR DESCRIPTION
**Pull request by jankratochvilcz**
## Summary

Attribute-only migration for `AssignProjectConfiguration` — adds `[MSBuildMultiThreadableTask]` to declare the task safe for multithreaded scheduling.

### Why attribute-only?

- No `Environment.CurrentDirectory`, `Path.GetFullPath`, `File.*`, or `ProcessStartInfo` usage
- All path resolution uses MSBuild's `GetMetadata(""FullPath"")` which returns engine-resolved absolute paths
- All mutable state is instance-scoped (platform mapping dicts, solution config cache)
- Only static method (`SetBuildInProjectAndReferenceOutputAssemblyMetadata`) is pure/stateless

No `IMultiThreadableTask` implementation needed. No behavioral change. No ChangeWave required.

Fixes #13613
Parent epic: #11834

### CODE REVIEW

File: src/Tasks/AssignProjectConfiguration.cs [hunk 1]
```diff
@@ -13,6 +13,7 @@
 
 namespace Microsoft.Build.Tasks
 {
+    [MSBuildMultiThreadableTask]
```
[the remark this answers is not in the document]
Comment (jankratochvilcz) [reply src/Tasks/AssignProjectConfiguration.cs:16]: Pushing back on this — a concurrency test is not justified for an attribute-only migration:

1. **Per-instance state**: MSBuild creates a fresh task instance per Execute() call (verified in the expert review on this PR). All mutable state (`_mappingsPopulated`, `_vcxToDefaultMap`, `_defaultToVcxMap`, base class `_cachedProjectReferencesByAbsolutePath`, `_solutionConfiguration`) is instance-scoped, so concurrent task instances cannot interfere by construction.

2. **No process-global state touched**: The expert review on this PR systematically verified no `Environment.CurrentDirectory`, no `Path.GetFullPath`, no `File.*`/`Directory.*`, no `ProcessStartInfo`, no static mutable state. `GetMetadata(\"FullPath\")` resolves through `FileUtilities.CurrentThreadWorkingDirectory` (an `AsyncLocal<string?>`) under the multithreaded driver, not `Environment.CurrentDirectory`.

3. **Pattern consistency**: ~27 other attribute-only migrations in the codebase (e.g., `CreateProperty`, `Message`, `AssignCulture`, `CombinePath`) ship without concurrency tests for the same reason — the attribute is a metadata declaration, not behavior. A concurrency test for an attribute-only migration would be exercising MSBuild's task instantiation model, not this task.

The concurrency-safety claim is a property of the absence of shared mutable state, which is an audit outcome, not a runtime invariant a test can falsify.

**Comment by OvesN** [comment]
@AR-May 
@JanProvaznik 
 Here we have this call chain: Execute() → new SolutionConfiguration(xmlString) → 
 FileUtilities.GetFullPathNoThrow(projectAbsolutePath).

 The projectAbsolutePath is read from the AbsolutePath attribute of <ProjectConfiguration> elements in
 the SolutionConfigurationContents XML (e.g. <ProjectConfiguration Project="{GUID}"
 AbsolutePath="c:\foo\A.csproj" ...>).

  Is it possible that this AbsolutePath attribute could ever contain a relative path? If a relative path
  ever slipped in, it would be a silent behavior change.

  If we expect this to always be an absolute path, should we add an assert (Path.IsPathRooted) in 
  SolutionConfiguration to make this contract explicit?

**Review by OvesN — APPROVED** [review]
Yeah, I think just attribute is enough to make this task thread safe **but only** if we honor the absolute-path contract in XML ProjectConfiguration.AbsolutePath attribute.

**Review by JanProvaznik — APPROVED** [review]

**Comment by OvesN** [comment]
> @AR-May @JanProvaznik Here we have this call chain: Execute() → new SolutionConfiguration(xmlString) → FileUtilities.GetFullPathNoThrow(projectAbsolutePath).
> 
> The projectAbsolutePath is read from the AbsolutePath attribute of elements in the SolutionConfigurationContents XML (e.g. <ProjectConfiguration Project="{GUID}" AbsolutePath="c:\foo\A.csproj" ...>).
> 
> Is it possible that this AbsolutePath attribute could ever contain a relative path? If a relative path ever slipped in, it would be a silent behavior change.
> 
> If we expect this to always be an absolute path, should we add an assert (Path.IsPathRooted) in SolutionConfiguration to make this contract explicit?

Rainer said that there is no need in assert. One source of this XML is us the second is VS, so we will trust them.
